### PR TITLE
Feature/impl

### PR DESCRIPTION
**Pull request by xRuBYN**

### CODE REVIEW

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 1]
```diff
@@ -2,68 +2,90 @@
...
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public class UserServiceImpl implements UserService {
 
     @Override
     public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().map(u -> u.getFirstName()).sorted((String u1, String u2) -> (u2.compareTo(u1))).collect(Collectors.toList());
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: instead 'u -> u.getFirstName()' use please method reference ClassName::methodName
for sorting you ca use Comparator.reverseOrder()

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 2]
```diff
@@ -2,68 +2,90 @@
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java:2]: could you please use this formatting for all lambda expressions, ex:
users.stream()
                .map(...
                .sorted(...
                .collect(Collectors.toList());

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 3]
```diff
@@ -2,68 +2,90 @@
...
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream().map(u -> u.getFirstName()).sorted((String u1, String u2) -> (u2.compareTo(u1))).collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        List<User> list = users.stream().sorted((User u1,User u2) ->(u1.getFirstName().compareTo(u2.getFirstName()))).sorted((User u1, User u2) -> u2.getAge().
+                compareTo(u1.getAge())).collect(Collectors.toList());
+        return list;
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: if you just return local variable 'list, you can remove it

for sorting you can use Comparator.comparing and for desc you can specify reversed()
ex: Comparator.comparing(method reference).reversed() 

if you have to sort by multiple fields you can use .thenComparing()

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 4]
```diff
@@ -2,68 +2,90 @@
...
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        List<List<Privilege>> list = users.stream().map(x -> x.getPrivileges()).collect(Collectors.toList());
+        List<Privilege> list1 = list.stream().flatMap(x -> x.stream().distinct()).distinct().collect(Collectors.toList());
+        return list1;
     }
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java:41]: image
and use please method reference where is possible

x -> x.stream().distinct() could be replaced with List::stream

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 5]
```diff
@@ -2,68 +2,90 @@
...
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        Optional<User> user = users.stream().filter(x->x.getAge() > age).filter(u -> u.getPrivileges().contains(Privilege.UPDATE)).findFirst();
+        return user;
     }
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java:48]: we don't need local variable user

instead filter(condition 1).filter(condition 2)
you can use filter(condition 1 && condition 2), and instead 'x', 'u' you could use `user`

Comment (xRuBYN) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java:48]: resolved

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 2 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java:2]: resolved

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 1 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: resolved

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 4 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java:41]: resolve

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 6]
```diff
@@ -2,68 +2,107 @@
...
+                .filter(user -> user.getAge() > age && user.getPrivileges().contains(Privilege.UPDATE))
+                .findFirst();
     }
 
     @Override
     public Map<Integer, List<User>> groupByCountOfPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(u -> Pair.of(u.getPrivileges().size(),u))
+                .collect(Collectors.groupingBy(Pair::getKey,Collectors.mapping(Pair::getValue,Collectors.toList())));
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]:
```suggestion
        users.stream()
                .collect(Collectors.groupingBy(user -> user.getPrivileges().size()))
```

Comment (xRuBYN) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: resolved

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 7]
```diff
@@ -2,68 +2,106 @@
...
+                        .map(privilege -> Pair.of(privilege,u)))
+                .collect(Collectors.groupingBy(Pair::getKey, Collectors.mapping(Pair::getValue, Collectors.toList())));
     }
 
     @Override
     public Map<String, Long> getNumberOfLastNames(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .map(u -> Pair.of(u.getLastName(), 1))
+                .collect(Collectors.groupingBy(Pair::getKey,Collectors.summingLong(Pair::getValue)));
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]:
```suggestion
        users.stream()
                .collect(Collectors.groupingBy(User::getLastName, Collectors.counting()));
```

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 8]
```diff
@@ -2,68 +2,106 @@
...
+                .orElse(-1);
     }
 
     @Override
     public Optional<String> getMostFrequentLastName(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        Map<String,Integer> map = users.stream()
+                .map(u-> Pair.of(u.getLastName(),1))
+                .collect(Collectors.groupingBy(Pair::getKey, Collectors.summingInt(Pair::getValue)));
+        if(map.values().stream().distinct().collect(Collectors.toList()).size() == 1) {
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: if(map.values().stream().distinct().count() == 1) {

File: src/main/java/com/endava/internship/service/UserServiceImpl.java [hunk 9]
```diff
@@ -2,68 +2,106 @@
...
+            return Optional.empty();
+        }
+        Optional<Map.Entry<String, Integer>> maxEntry = map.entrySet()
+                .stream()
+                .max(Comparator.comparing(Map.Entry::getValue));
```
Comment (meedvv379) [inline src/main/java/com/endava/internship/service/UserServiceImpl.java]: .max(Map.Entry.comparingByValue());

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 7 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: resolved

File: src/main/java/com/endava/internship/service/UserServiceImpl.java (hunk 8 above)
Comment (xRuBYN) [reply src/main/java/com/endava/internship/service/UserServiceImpl.java]: resolved

**Review by meedvv379 — APPROVED** [review]

**Pull request merged**